1976-2 Don't Delete Users

dovek opened the pull request:
### Status
**IN DEVELOPMENT**

### Description
As an instructor, if I delete a student, they should be removed from having an 'active' course membership, but the course membership should continue to exist, and they should be retrievable - including all grades, earned badges, submissions, etc.

This is valuable for both data purposes (knowing the history of what happened), and managing user errors.

### Related PRs
N/A


### Todos
- [x] Add Tests
- [x] Finish Development


### Deploy Notes
Will need to run db:migrate

### Gem dependencies
N/A

### Migrations
YES

### Steps to Test or Reproduce
TO BE COMPLETED

### Impacted Areas in Application
List general components of the application that this PR will affect:

* Users
* Course Memberships

======================
Closes #1976

chcholman commented:
@dovek This is looking fantastic! I did find a couple of additional places we need to filter these students out. Those are: 

- [ ] Gradebook (the deactivated user shows as a row, and should not appear here at all)
- [ ] Awarded Badge Chart (the deactivated user shows as a row, and should not appear here at all)
- [ ] Multiplier Choices chart (the deactivated user shows as a row, and should not appear here at all)
- [ ] Grading Status page (deactivated users ungraded submissions, resubmissions, and unreleased grades all show as individual rows here)
- [ ] Badges index (deactivated users show in the total of badges earned for each badge, and should be removed from the count)
- [ ] Badges show page (deactivated users show as a row, and should not appear here at all)
- [ ] Badges Quick Award/mass edit (deactivated users show as a row, and should not appear here at all)
- [ ] Badges New page (deactivated users show in the dropdown list of students to award to, and should not appear here)
- [ ] Assignments index (deactivated users show in the counts for Grades and Submissions, and should not be included here)
- [ ] Assignments show page (deactivated users show as a row, and should be filtered out from here)
- [ ] Assignments Quick Grade/mass_edit (deactivated users show as a row for you to enter grades, and should not appear) 
- [ ] Teams index (deactivated users contribute to the average earned score, and should not)
- [ ] Teams show page (deactivated users appear as a row, and should be filtered out)

dovek commented:
@chcholman Regarding the SQL error I was getting earlier, I added it to the file app/models/submission.rb. The issue that I'm getting is it looks like ActiveRecord is duplicating the joins. When I go into the rails console and run Submission.resubmitted.to_sql to get the SQL I can see that the from clause is a little strange:

```
 FROM "submissions" 
 LEFT OUTER JOIN "assignments" ON "assignments"."id" = "submissions"."assignment_id" 
 LEFT OUTER JOIN "grades" ON "grades"."submission_id" = "submissions"."id" 
 LEFT OUTER JOIN "users" ON "users"."id" = "submissions"."student_id" 
 LEFT OUTER JOIN "groups" ON "groups"."id" = "submissions"."group_id" 
 LEFT OUTER JOIN "submission_files" ON "submission_files"."submission_id" = "submissions"."id" 
 INNER JOIN course_memberships ON course_memberships.course_id = submissions.course_id AND course_memberships.user_id = submissions.student_id 
 INNER JOIN course_memberships ON course_memberships.course_id = submissions.course_id AND course_memberships.user_id = submissions.student_id 
 LEFT OUTER JOIN grades ON grades.submission_id = submissions.id 
 LEFT OUTER JOIN assignments ON assignments.id = submissions.assignment_id 
 WHERE "submissions"."course_id" = $1 AND (course_memberships.active = true AND submitted_at is not null) 
 AND (course_memberships.active = true) 
 AND (grades.status = 'Released' 
 OR (grades.status = 'Graded' AND NOT assignments.release_necessary)) 
 AND (grades.graded_at < submitted_at)
```

In addition to this, I see that I've introduced some errors in the specs that I will take a look at this evening/tomorrow. If you have any comments, questions, or concerns please let me know.

chcholman commented on app/assets/javascripts/dynatable_student_index.js:
```diff
@@ -8,21 +8,30 @@ if($('#student-index-table').length > 0 ) {
+      dynatable.queries.functions['deactivated'] = function(record) {
+        return record.active == "false";
+      };
+      dynatable.queries.functions['deactivated'] = function(record) {
```
I believe this is duplicated, and one can be removed?

chcholman commented on app/models/badge.rb:
```diff
@@ -40,4 +40,12 @@ def available_for_student?(student)
     earned_badges.where(student_id: student.id).count < 1
   end
+
+  def earned_badge_total_points_for_student(student)
```
Did these two methods get moved over to the API and maybe came back here during a rebase?

chcholman replied:
(If not, check with @acousticrobot about whether or not they should be moved over in a separate issue)

chcholman commented on app/models/concerns/analytics/assignment_analytics.rb:
```diff
@@ -4,25 +4,25 @@ module AssignmentAnalytics
     # Average of above-zero grades for an assignment
     def earned_average
-      grades.graded_or_released.where("score > 0").average(:score).to_i
+      grades.for_active_students.graded_or_released.where("grades.score > 0").average(:score).to_i
```
Was it throwing an error without the addition of `grades`? I'm curious why that got added

chcholman replied:
For consistency reasons, let's move the `for_active_students` to after `graded_or_released`

dovek replied:
@chcholman yes, it was throwing an error in the spec:
```
     ActiveRecord::StatementInvalid:
       PG::AmbiguousColumn: ERROR:  column reference "score" is ambiguous
       LINE 1: ...sed') AND (course_memberships.active = true) AND (score > 0)
```

chcholman commented on app/models/concerns/analytics/course_analytics.rb:
```diff
@@ -4,27 +4,27 @@ module CourseAnalytics
 
     def scores
-      scorable_memberships.pluck(:score).sort
+      course_memberships.being_graded.pluck(:score).sort
```
I believe this should go back to being `scorable_memberships`

chcholman commented on app/models/concerns/analytics/course_analytics.rb:
```diff
@@ -4,27 +4,27 @@ module CourseAnalytics
 
     def average_score
-      scorable_memberships.average(:score)
+      course_memberships.being_graded.average(:score)
```
same as above

chcholman commented on app/models/concerns/analytics/course_analytics.rb:
```diff
@@ -4,27 +4,27 @@ module CourseAnalytics
 
     def high_score
-      scorable_memberships.maximum(:score)
+      course_memberships.being_graded.maximum(:score)
```
same as above

chcholman commented on app/models/course.rb:
```diff
@@ -156,6 +156,18 @@ def assignment_weight_spent_for_student(student)
   end
 
+  def student_count
+    course_memberships.where(role: "student", active: true).count
```
All three of these have been moved over to the `course_analytics` api and can be removed (depending, you may need to update the references to them)

chcholman commented on app/views/assignments/_index_staff.haml:
```diff
@@ -55,7 +55,7 @@
                   %td= "Yes" if assignment.grade_with_rubric?
                   %td= assignment.submissions.submitted.count if assignment.accepts_submissions? && assignment.submissions.present?
-                  %td= assignment.grades.graded_or_released.count if assignment.grades.instructor_modified.any?
+                  %td= assignment.grades.graded_or_released.for_active_students.count if assignment.grades.instructor_modified.any?
```
Let's add a `for_active_students` to the if check, otherwise you might have grades that were edited for non-active students, and you'll see a 0 here which will be confusing for the user

chcholman commented:
@dovek It's looking really good Kyle! Some minor tweaks, most of which I think are probably due to the rebasing. Let me know if I can help in any way!

chcholman commented:
Looks great, thank you @dovek!
